### ISSUE TITLE
Parenthesized expressions are not handled correctly by the debug hover

### ISSUE THREAD

**Issue by FMorschel** [Dart-Code/Dart-Code#5362]
**Describe the bug**
Parenthesized expressions are not handled correctly by the debug hover.

**To Reproduce**

```dart
void main() {
  const n = 1;
  n.asDouble;
  (n).asDouble; //Breakpoint here
}

extension on int {
  double get asDouble => toDouble();
}
```

If you now place a break point on the indicated line and hover `asDouble` there and the line above, you'll see the different results as the line above does show the value.

**Expected behavior**
Both to show the value.

**Please complete the following information:**

<details>

<summary><strong>Workspace Environment</strong></summary>

```text
Dart Code extension: 3.103.20241202
Flutter extension: 3.103.20241202 (activated)

App: Visual Studio Code
App Host: desktop
Version: win 1.95.3

Workspace type: Dart (LSP)
Workspace name: bug

Dart (3.7.0-edge.579de4e2d38acb025be5135ef6ce2b9104e208d9): C:\Users\felip_0vh5fa6\dart
Flutter (3.24.5): C:\Users\felip_0vh5fa6\.puro\envs\stable\flutter (No device)

```

</details>

<details>

<summary><strong>Output from 'dart info'</strong></summary>

`C:\Users\felip_0vh5fa6\dart\bin\dart.exe info`

If providing this information as part of reporting a bug, please review the information
below to ensure it only contains things you're comfortable posting publicly.

#### General info

- Dart 3.7.0-edge.579de4e2d38acb025be5135ef6ce2b9104e208d9 (main) (Mon Dec 2 23:15:39 2024 +0000) on "windows_x64"
- on windows / "Windows 11 Pro" 10.0 (Build 22631)
- locale is pt-BR

#### Project info

- sdk constraint: '^3.7.0-edge.d75c493e27f9bf2a89043860132462be665065c2'
- dependencies: 
- dev_dependencies: lints, test

#### Process info

| Memory | CPU | Elapsed time | Command line |
| -----: | --: | -----------: | ------------ |
|   0 MB |  -- |              | dart.exe     |
|   0 MB |  -- |              | dart.exe     |
|   0 MB |  -- |              | dart.exe     |
|   0 MB |  -- |              | dart.exe     |
|   0 MB |  -- |              | dart.exe     |

</details>

**Comment by DanTup** [comment]
@bkonyi the issue here is that when you hover something in the editor, VS Code has to "guess" what to expand the selection to to get the expression to evaluate. It also fails in some other ways (for example because of scoping).

We do also have a line/col for where the user hovered (in the debug adapter, not in the VS Code extension). Is there an existing way we could evaluate "the expression at a given line/col" via the VM Service (and have it automatically expand to the nearest valid expression)?

**Comment by bkonyi** [comment]
I don't think we currently have any support along those lines. Maybe Derek (I can't tag him here) will have some ideas since he's been working on expression evaluation related stuff.

**Comment by DanTup** [comment]
@derekxu16 any thoughts on this?

Currently evaluation when you hover over something is done by VS Code "guessing" what the correct expression is. It's sometimes wrong because it doesn't know what's valid, but also it can't take into account scope... For example:

```dart
var aaa = 1;

f() {
  var aaa = 2;
  debugger();
}
```

If you hit `debugger();` here and hover over the first variable `aaa`, we'd show the wrong value. If we were passing the line/col to the VM, it could (presumably) solve both of the issues in one?

**Comment by derekxu16** [comment]
> If you hit `debugger();` here and hover over the first variable `aaa`, we'd show the wrong value.

I believe that this issue can be resolved by calling [`getStack`](https://github.com/dart-lang/sdk/blob/main/runtime/vm/service/service.md#getStack), walking up the stack until finding a frame that is running a function whose `SourceLocation` contains the location of the hovered expression, and then using [`evaluteInFrame`](https://github.com/dart-lang/sdk/blob/main/runtime/vm/service/service.md#evaluateinframe) to evaluate the hovered expression in that frame. @bkonyi, do you think we should add a DDS method that does all this, or should we just leave it to the clients to implement what I've described?

> Is there an existing way we could evaluate "the expression at a given line/col" via the VM Service (and have it automatically expand to the nearest valid expression)?

I need to think about this issue some more. I'll get back to you after I do.

**Comment by DanTup** [comment]
> @bkonyi, do you think we should add a DDS method that does all this, or should we just leave it to the clients to implement what I've described?

I think that would be a nice solution. Having clients do it (although for VS Code the client is actually the DAP code also in the SDK) could lead to differences (or some not implementing it). I think a single central implementation (close to the code it relies on) would be better.

> > Is there an existing way we could evaluate "the expression at a given line/col" via the VM Service (and have it automatically expand to the nearest valid expression)?
> 
> I need to think about this issue some more. I'll get back to you after I do.

If we were to have an API for the above, I think this would make sense to be part of it. Eg., the API is an `evaluateExpressionAtPosition` API that just takes a line/col (or tokenpos or something) and evaluates whatever makes sense for the token at that location. (I don't know if what "makes sense" is obvious/unambigious, but again, having it central rather than left to the clients I think would be a better option).

**Comment by bkonyi** [comment]
> > If you hit `debugger();` here and hover over the first variable `aaa`, we'd show the wrong value.
> 
> I believe that this issue can be resolved by calling [`getStack`](https://github.com/dart-lang/sdk/blob/main/runtime/vm/service/service.md#getStack), walking up the stack until finding a frame that is running a function whose `SourceLocation` contains the location of the hovered expression, and then using [`evaluteInFrame`](https://github.com/dart-lang/sdk/blob/main/runtime/vm/service/service.md#evaluateinframe) to evaluate the hovered expression in that frame. @bkonyi, do you think we should add a DDS method that does all this, or should we just leave it to the clients to implement what I've described?

We should definitely provided a shared implementation. Adding it to DDS SGTM in principle assuming we have a clean enough API.

**Comment by DanTup** [comment]
I've opened https://github.com/dart-lang/sdk/issues/59737 in the SDK about this.

**Comment by FMorschel** [comment]
Not 100% sure but this might be another case related to this:

```dart
var list = [1, 2, 3]; // Hover `list`

void main() {
  var other = [ // Breakpoint here
    ...list, // Hover `list`
  ];
}
```

**Comment by FMorschel** [comment]
One more case:

```dart
int? variable = 0;

void main() {
  var other = [ // Breakpoint here
    variable!.asDouble, // Hover `asDouble`
    variable?.asInt, // Hover `asInt`
  ];
}

extension on num {
  double get asDouble => toDouble();
  int get asInt => toInt();
}
```

You can see here that the `!.` sequence also makes this fail.

**Comment by FMorschel** [comment]
Here this is evaluated but it could be improved when we have https://github.com/dart-lang/sdk/issues/59737.

```dart
int? variable = 0;

void main() {
  variable..isEven;
}
```

When you hover `isEven` it still evaluates the whole expression (so it results `0` as `variable`). Maybe we could then evaluate `isEven` in this case when the cursor is over it and not `variable`.
